### PR TITLE
feat: reduce build times for most libraries

### PR DESCRIPTION
**Pull request by dbolduc**
Part of the work for #14108 

Generate conglomerate source files for gRPC-based full GAPICs. In our testing we saw a ~3x speed up in build times of just the library. Note that the cost of building dependencies, or common components of `google-cloud-cpp` are not affected by this change.

The changes are split into multiple commits for reviewability.

Unaffected libraries (to be addressed later):
- `bigtable`
- `compute`
- `pubsub`
- `spanner`
- `sql`
- `storage`

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14215)

### CODE REVIEW

File: generator/internal/sources_generator.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * This conglomerate source is essentially a concatenation of the other source
+ * files emitted by the generator, for a service.
+ *
+ * Every translation unit (TU) that transitively includes a protobuf header
+ * needs to recompile that header. To minimize build times, it is best to have
+ * just one TU.
+ *
+ * See go/cloud-cxx:reducing-build-times for further discussion on the design,
+ * plus some experimental results.
```
Comment (devbww) [inline generator/internal/sources_generator.h:40]: Don't name internal documents?

File: generator/internal/sources_generator.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ */
+class SourcesGenerator : public ServiceCodeGenerator {
+ public:
+  SourcesGenerator(
+      google::protobuf::ServiceDescriptor const* service_descriptor,
+      VarsDictionary service_vars,
+      std::map<std::string, VarsDictionary> service_method_vars,
+      google::protobuf::compiler::GeneratorContext* context,
+      std::vector<std::string> sources);
```
Comment (devbww) [inline generator/internal/sources_generator.h:49]: `#include <vector>`

File: generator/internal/make_generators.cc [hunk 3]
```diff
@@ -155,6 +160,32 @@ std::vector<std::unique_ptr<GeneratorInterface>> MakeGenerators(
...
+    if (generate_grpc_transport) {
+      sources.push_back(service_vars["stub_factory_cc_path"]);
+      sources.push_back(service_vars["auth_cc_path"]);
+      sources.push_back(service_vars["logging_cc_path"]);
+      sources.push_back(service_vars["metadata_cc_path"]);
+      sources.push_back(service_vars["stub_cc_path"]);
+      sources.push_back(service_vars["tracing_stub_cc_path"]);
+    }
+
+    std::sort(sources.begin(), sources.end());
```
Comment (devbww) [inline generator/internal/make_generators.cc:184]: `#include <algorithm>`

File: bazel/gapic.bzl [hunk 4]
```diff
@@ -26,16 +26,18 @@ def cc_gapic_library(name, service_dirs = [], googleapis_deps = []):
             this library depends on.
     """
 
-    src_dirs = service_dirs + [d + "internal/" for d in service_dirs]
+    code_glob = [d + i + f for d in service_dirs for i in ["", "internal/"] for f in ["*.h", "*.cc"]]
+
+    sources_glob = [d + "internal/*_sources.cc" for d in service_dirs] + [d + "internal/streaming.cc" for d in service_dirs]
```
Comment (devbww) [inline bazel/gapic.bzl:31]: Some judicious line breaking would be useful here.

File: generator/internal/sources_generator.h (hunk 1 above)
Comment (alevenberg) [reply generator/internal/sources_generator.h:40]: We do this in some other places in the code.

Comment (dbolduc) [reply generator/internal/sources_generator.h:40]: OSS code and internal documents don't exactly mix. I would always avoid linking internal documents in public documentation.

This documentation is internal though. It is for us. I like having a link to why we thought this was a good idea, even if some of us may not be able to access it tomorrow.

File: .clang-tidy [hunk 5]
```diff
@@ -61,6 +61,9 @@
 #     Almost any expression of the form `2 * variable` or `long x = a_int * b_int;`
 #     generates an error.
 #
+#  -bugprone-suspicious-include: We `#include "*.cc"` files as a way to squash
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply .clang-tidy:64]: I like what you are proposing better than suppressing `bugprone-suspicious-include` globally.

Let me address this in a follow up PR, though.

I will need to either do some plumbing in the generator:
https://github.com/googleapis/google-cloud-cpp/blob/e7d3944b5a85f80351f7db13e4ef4bde89d55469/generator/internal/codegen_utils.cc#L207-L210

Or just write out the text myself.
https://github.com/googleapis/google-cloud-cpp/blob/0bba30dfc14854e970516e997a7445d8b296ff90/generator/internal/sources_generator.cc#L40

File: generator/internal/sources_generator.h (hunk 2 above)
Comment (dbolduc) [reply generator/internal/sources_generator.h:50]: Done.

File: generator/internal/make_generators.cc (hunk 3 above)
Comment (dbolduc) [reply generator/internal/make_generators.cc:185]: Done.

File: bazel/gapic.bzl (hunk 4 above)
Comment (dbolduc) [reply bazel/gapic.bzl]: I tried.

File: .clang-tidy (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbolduc) [reply .clang-tidy:64]: I added a line item to the parent issue to track the cleanup.

[the remark this answers is not in the document]
Comment (dbolduc) [reply .clang-tidy]: > Let me address this in a follow up PR

Nvm. Fixed here.

@devbww asked if we could do `NOLINT` blocks, and the answer is yes, now that we are on clang-tidy >= 14.

PTAL

**Review by coryan — APPROVED** [review]

**Pull request merged**